File: frontend/src/types/pool.ts
import type { UserNeedProfile } from "./onboarding"
import type { PreferenceSnapshot } from "./preferences"

export interface TimeWindow {
  start: string
  end: string
}

export interface PoolRequest {
  user_id: string
  city: string
  date: string
  time_window: TimeWindow
  persona_tags: string[]
  pace_style?: string
  party?: string
  budget_per_person?: number
  free_text?: string
  need_profile?: UserNeedProfile
  preference_snapshot?: PreferenceSnapshot
  origin_latitude?: number
  origin_longitude?: number
  radius_meters?: number
}

export interface EvidenceSnippet {
  doc_id: string
  source_type: "poi_profile" | "ugc_review" | "fts" | "feature_bucket" | string
  text: string
  score: number
}

export interface PoiInPool {
  id: string
  name: string
  category: string
  latitude: number
  longitude: number
  rating: number
  price_per_person?: number | null
  cover_image?: string | null
  distance_meters?: number | null
  why_recommend: string
  highlight_quote?: string | null
  keywords: string[]
  estimated_queue_min?: number | null
  suitable_score: number
  score_breakdown: Record<string, number>
  retrieval_score?: number | null
  retrieval_provenance: string[]
  evidence_snippets: EvidenceSnippet[]
}

export interface PoolCategory {
  name: string
  description: string
  pois: PoiInPool[]
}

export interface PoolMeta {
  total_count: number
  generated_at: string
  user_persona_summary: string
  data_warning?: string | null
}

export interface PoolResponse {
  pool_id: string
  categories: PoolCategory[]
  default_selected_ids: string[]
  meta: PoolMeta
}

File: backend/app/schemas/pool.py
from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field, field_validator

from app.schemas.onboarding import UserNeedProfile
from app.schemas.preferences import PreferenceSnapshot
from app.schemas.user_memory import UserFacts


class TimeWindow(BaseModel):
    start: str
    end: str

    @field_validator("start", "end")
    @classmethod
    def validate_hhmm(cls, value: str) -> str:
        parts = value.split(":")
        if len(parts) != 2:
            raise ValueError("time must be HH:MM")
        hour, minute = int(parts[0]), int(parts[1])
        if hour < 0 or hour > 23 or minute < 0 or minute > 59:
            raise ValueError("time must be HH:MM")
        return value


class PoolRequest(BaseModel):
    user_id: str
    city: str
    date: str = "2026-05-02"
    time_window: TimeWindow = Field(default_factory=lambda: TimeWindow(start="13:00", end="21:00"))
    persona_tags: list[str] = Field(default_factory=list)
    pace_style: Optional[str] = None
    party: Optional[str] = None
    budget_per_person: Optional[int] = None
    free_text: Optional[str] = None
    need_profile: Optional[UserNeedProfile] = None
    preference_snapshot: Optional[PreferenceSnapshot] = None
    user_facts: Optional[UserFacts] = None
    ugc_hits: list[dict[str, Any]] = Field(default_factory=list)


class PoiInPool(BaseModel):
    id: str
    name: str
    category: str
    rating: float
    price_per_person: Optional[int]
    cover_image: Optional[str]
    distance_meters: Optional[int]
    why_recommend: str
    highlight_quote: Optional[str]
    keywords: list[str]
    estimated_queue_min: Optional[int]
    suitable_score: float
    score_breakdown: dict[str, float] = Field(default_factory=dict)


class PoolCategory(BaseModel):
    name: str
    description: str
    pois: list[PoiInPool]


class PoolMeta(BaseModel):
    total_count: int
    generated_at: datetime
    user_persona_summary: str


class PoolResponse(BaseModel):
    pool_id: str
    categories: list[PoolCategory]
    default_selected_ids: list[str]
    meta: PoolMeta
